File: performance-comparison/benchmarks/haskell-project/src/Nbody.hs
-- nbody -- LCG-initialized bodies over mutable unboxed doubles, pairwise
-- forces, kinetic-energy checksum at %.4f.  Init casts the LCG's top 32
-- bits through SIGNED 64-bit like the C/Turmeric column.
{-# LANGUAGE BangPatterns #-}
import System.Environment (getArgs)
import Text.Printf (printf)
import Foreign.Marshal.Alloc (mallocBytes, free)
import Foreign.Ptr (Ptr)
import Foreign.Storable (peekElemOff, pokeElemOff)
import Data.Word (Word64)
import Data.Int (Int64)
lcgA, lcgC :: Word64
lcgA = 6364136223846793005
lcgC = 1442695040888963407
-- layout: 7 doubles per body: x y z vx vy vz mass
main :: IO ()
main = do
  args <- getArgs
  let (nBodies, steps) = case args of
        (a:b:_) -> (read a, read b)
        (a:_)   -> (read a, 1000)
        _       -> (5, 1000) :: (Int, Int)
  p <- mallocBytes (nBodies * 7 * 8) :: IO (Ptr Double)
  let lcgF s = fromIntegral (fromIntegral (s `div` 4294967296) :: Int64) / 1e8
      initB !i !st
        | i >= nBodies = return ()
        | otherwise = do
            let s1 = st * lcgA + lcgC
                s2 = s1 * lcgA + lcgC
                s3 = s2 * lcgA + lcgC
            pokeElemOff p (i * 7 + 0) (lcgF s1)
            pokeElemOff p (i * 7 + 1) (lcgF s2)
            pokeElemOff p (i * 7 + 2) (lcgF s3)
            pokeElemOff p (i * 7 + 3) 0.0
            pokeElemOff p (i * 7 + 4) 0.0
            pokeElemOff p (i * 7 + 5) 0.0
            pokeElemOff p (i * 7 + 6) (1.0 + fromIntegral (i `mod` 5) * 0.5)
            initB (i + 1) s3
  initB 0 42
  let pair !i !j
        | j >= nBodies = return ()
        | otherwise = do
            xi <- peekElemOff p (i*7+0); yi <- peekElemOff p (i*7+1); zi <- peekElemOff p (i*7+2)
            xj <- peekElemOff p (j*7+0); yj <- peekElemOff p (j*7+1); zj <- peekElemOff p (j*7+2)
            mi <- peekElemOff p (i*7+6); mj <- peekElemOff p (j*7+6)
            let dx = xj - xi; dy = yj - yi; dz = zj - zi
                dist = sqrt (dx*dx + dy*dy + dz*dz) + 1e-10
                f = mi * mj / (dist * dist * dist)
            vxi <- peekElemOff p (i*7+3); pokeElemOff p (i*7+3) (vxi + f*dx)
            vyi <- peekElemOff p (i*7+4); pokeElemOff p (i*7+4) (vyi + f*dy)
            vzi <- peekElemOff p (i*7+5); pokeElemOff p (i*7+5) (vzi + f*dz)
            vxj <- peekElemOff p (j*7+3); pokeElemOff p (j*7+3) (vxj - f*dx)
            vyj <- peekElemOff p (j*7+4); pokeElemOff p (j*7+4) (vyj - f*dy)
            vzj <- peekElemOff p (j*7+5); pokeElemOff p (j*7+5) (vzj - f*dz)
            pair i (j + 1)
      move !i
        | i >= nBodies = return ()
        | otherwise = do
            x <- peekElemOff p (i*7+0); vx <- peekElemOff p (i*7+3)
            y <- peekElemOff p (i*7+1); vy <- peekElemOff p (i*7+4)
            z <- peekElemOff p (i*7+2); vz <- peekElemOff p (i*7+5)
            pokeElemOff p (i*7+0) (x + vx)
            pokeElemOff p (i*7+1) (y + vy)
            pokeElemOff p (i*7+2) (z + vz)
            move (i + 1)
      step !s
        | s >= steps = return ()
        | otherwise = do
            sequence_ [pair i (i + 1) | i <- [0 .. nBodies - 1]]
            move 0
            step (s + 1)
  step 0
  let ke !i !acc
        | i >= nBodies = return acc
        | otherwise = do
            vx <- peekElemOff p (i*7+3); vy <- peekElemOff p (i*7+4)
            vz <- peekElemOff p (i*7+5); m <- peekElemOff p (i*7+6)
            ke (i + 1) (acc + 0.5 * m * (vx*vx + vy*vy + vz*vz))
  e <- ke 0 0.0
  free p
  printf "%.4f\n" e

File: performance-comparison/benchmarks/rust-workspace/src/bin/nbody.rs
// nbody -- LCG-initialized bodies, pairwise force simulation, kinetic
// energy checksum at %.4f. Init casts the LCG's top 32 bits through
// signed 64-bit, like the C/Turmeric/Haskell columns.
use std::env;

const LCG_A: u64 = 6364136223846793005;
const LCG_C: u64 = 1442695040888963407;

#[derive(Clone, Copy)]
struct Body {
    x: f64,
    y: f64,
    z: f64,
    vx: f64,
    vy: f64,
    vz: f64,
    mass: f64,
}

fn lcg_f(s: u64) -> f64 {
    ((s >> 32) as i64) as f64 / 1e8
}

fn run_nbody(n_bodies: i64, steps: i64) -> f64 {
    let n = n_bodies as usize;
    let mut bodies: Vec<Body> = Vec::with_capacity(n);
    let mut state: u64 = 42;
    for i in 0..n {
        state = state.wrapping_mul(LCG_A).wrapping_add(LCG_C);
        let x = lcg_f(state);
        state = state.wrapping_mul(LCG_A).wrapping_add(LCG_C);
        let y = lcg_f(state);
        state = state.wrapping_mul(LCG_A).wrapping_add(LCG_C);
        let z = lcg_f(state);
        bodies.push(Body {
            x,
            y,
            z,
            vx: 0.0,
            vy: 0.0,
            vz: 0.0,
            mass: 1.0 + (i % 5) as f64 * 0.5,
        });
    }

    for _ in 0..steps {
        for i in 0..n {
            for j in (i + 1)..n {
                let dx = bodies[j].x - bodies[i].x;
                let dy = bodies[j].y - bodies[i].y;
                let dz = bodies[j].z - bodies[i].z;
                let dist = (dx * dx + dy * dy + dz * dz).sqrt() + 1e-10;
                let f = bodies[i].mass * bodies[j].mass / (dist * dist * dist);
                bodies[i].vx += f * dx;
                bodies[i].vy += f * dy;
                bodies[i].vz += f * dz;
                bodies[j].vx -= f * dx;
                bodies[j].vy -= f * dy;
                bodies[j].vz -= f * dz;
            }
        }
        for b in bodies.iter_mut() {
            b.x += b.vx;
            b.y += b.vy;
            b.z += b.vz;
        }
    }

    let mut ke = 0.0;
    for b in bodies.iter() {
        ke += 0.5 * b.mass * (b.vx * b.vx + b.vy * b.vy + b.vz * b.vz);
    }
    ke
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n_bodies: i64 = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(5);
    let steps: i64 = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(1000);
    println!("{:.4}", run_nbody(n_bodies, steps));
}
